src/browser.ts

```diff
@@ -1,41 +1,95 @@
+}
+
+function parsePathname(paths: PathConfigs<BrowserState>, location?: Location): Partial<BrowserState> {
+  return locationToPathConfig(paths, location).deserialize(locationToPathname(location))!;
+}
+
 // export is for tests only
 // TODO(gio): `location` here is optional just because (I think)
 // it can be undefined in some tests using `memoryHistory`
-export function parseBrowserState(location?: Location): BrowserState {
-  const view = location ? trim(location.pathname, ' /') : '';
+export function parseBrowserState(paths: PathConfigs<BrowserState>, location?: Location): BrowserState {
+  const pathState = parsePathname(paths, location);
   const s = location ? qs.parse(trim(location.search, ' ?')) : {};
-  return { ...s, view };
+  return { ...s, ...pathState };
+}
+
+function cleanupPathParams(
+  s: BrowserState,
+  currentPath: PathConfig<BrowserState>,
+  nextPath: PathConfig<BrowserState>
+): BrowserState {
+  const currentPathParamKeys = Object.keys(currentPath.pick(s));
+  const nextPathParamKeys = Object.keys(nextPath.pick(s));
+  return omit(s, difference(currentPathParamKeys, nextPathParamKeys));
+}
+
+function browserStateToPathConfig(s: BrowserState, paths: PathConfigs<BrowserState>): PathConfig<BrowserState> {
+  return find(paths, p => p.is(s)) || defaultNoPath;
 }
 
 // export is for tests only
-export function serializeBrowserState(s: BrowserState): Partial<Location> {
-  const pathname = `/${encodeURIComponent(s.view)}`;
-  const search = `?${qs.stringify(omit(s, ['view']))}`; // will encode URI by default
+export function serializeBrowserState(s: BrowserState, paths: PathConfigs<BrowserState>): Partial<Location> {
+  const path = browserStateToPathConfig(s, paths);
+  const pathState = path.pick(s);
+  const pathname = `/${path.serialize(mapValues(pathState, encodeURIComponent))}`;
+  const search = `?${qs.stringify(omit(s, Object.keys(pathState)))}`; // will encode URI by default
   return { pathname, search };
 }
 
-export default function(history: History = createBrowserHistory()) {
+export default function(paths: PathConfigs<BrowserState> = [], history: History = createBrowserHistory()) {
+  let currentPath = locationToPathConfig(paths, history.location);
+
   function syncToBrowser(s: BrowserState, push: boolean = true): void {
-    (push ? history.push : history.replace)(serializeBrowserState(s));
+    (push ? history.push : history.replace)(serializeBrowserState(s, paths));
   }
 
   function onBrowserChange(callback: (s: BrowserState, action: Action) => void): void {
     history.listen((location, action) => {
-      callback(parseBrowserState(location), action);
+      currentPath = locationToPathConfig(paths, location);
+      callback(parseBrowserState(paths, location), action);
     });
-    callback(parseBrowserState(history.location), 'PUSH');
+    callback(parseBrowserState(paths, history.location), 'PUSH');
+  }
+
+  function dryRunBrowserTransition<S extends BrowserState>(s: S): S {
```

Comment:
oh ok, it's used here. I guess this function is needed to know if the transition will change the state **before** actually running the transition?

Reply:
We always assumed *some* state keys (path params) would just go away when transitioning to a different path via the url.

Previously this was implemented always waiting for the browser and giving url the full control of path params (we marked a subset of state keys as "path params", react-router would give us the new path params, we trusted it and deleted every other "path param" before setting the new ones, if any).

Now we are never waiting for the browser to push the updated state (aka: the URL is always updated *after* a transition, except of course for when the change is actually initiated via browser controls).

This means that to maintain this semantics, we still have to threat some keys as special keys (hence the need for `pick` function in the `PathConfig` type) that should be automatically removed when exiting a certain "path". Also, this should be performed *before* pushing the updated state to the app (aka before `transition`).

I don't like this for the long term. I'd prefer having the user explicitly doing this, aided by types (making sure the new state doesn't have keys in eccess)
For now, this was simpler and a easier upgrade path